### DISCUSSION TITLE
Inserting New Tables into DB for Live Scoreboard

### DISCUSSION THREAD

**Discussion by RylanRCox** [PrairieLearn/PrairieLearn#8141] [Dev]
I'm a part of a capstone project that is attempting to create a live scoreboard for in-class interactivity. We are currently running into issues with setting up the Docker environment locally and inputting our own tables into the DB. Any help would be appreciated.

We attempted to follow this section of the dev guide which seems to only be for DB adjustments when we are hoping to insert whole new tables.
![image](https://github.com/PrairieLearn/PrairieLearn/assets/71582228/9aa05415-05cc-4bb4-876d-34314d690e88)

**Comment by eliotwrobson** [comment]
This sounds awesome! See related issue: https://github.com/PrairieLearn/PrairieLearn/issues/1151

I'm not very knowledgeable about the DB, but I've messed with the docker dev process some. What issues are you running into specifically @RylanRCox ?

**Reply by RylanRCox** [reply]
So the issue we're running into is none of the changes we make to any of the code are reflected when we pull up localhost:3000. We've tried taking down the docker container and restarting it, and we've tried loading from the disk. Neither of these has worked.
The commands we're running to start the container are:

docker run -it --rm -p 3000:3000 -w /PrairieLearn -v ${PWD}:/PrairieLearn prairielearn/prairielearn /bin/bash

Then in the container, we run:

make deps

make dev

Additionally, when we run make dev, it takes about 15 to 20 minutes for yarn dev to start up, is there any way to speed up this process?

Any help from this point on would be very appreciated. Thank you in advance.

**Reply by eliotwrobson** [reply]
> None of the changes we make to any of the code are reflected when we pull up localhost:3000.

This can be tricky to get working sometimes. If you don't set the start command quite right, you could just be running the version of PL in the container and not the code from the repo you've checked out. A container restart shouldn't be necessary though, you should be able to get away with just running `make build` in the container, then `make dev`.

> Additionally, when we run make dev, it takes about 15 to 20 minutes for yarn dev to start up, is there any way to speed up this process?

Which operating system are you developing on? I ask this because running things out of a docker container is known for being slow, and using docker on Windows uses WSL, which makes things even slower as well. If you're on Windows, make sure to have WSL2 installed (as that will speed things up over WSL), but if you're planning on spending substantial amounts of time on this project, I would _highly_ recommend working on a UNIX based machine (Linux or Mac, setting up a dual boot if you have to). To speed up the `yarn dev` time, you can even develop natively on your system outside of a container (see instructions [here](https://github.com/PrairieLearn/PrairieLearn/blob/master/docs/installingNative.md)). This can be trickier to set up, but will substantially speed up your development workflow.

> Any help from this point on would be very appreciated. Thank you in advance.

You are very welcome! I know we've been steering people more towards posting on GitHub discussions, but for more tech-support type help where it can be useful to interact directly, check out the [pl-help](https://app.slack.com/client/T0PHEMPRV/C266KEH9A) channel on slack.

**Reply by RylanRCox** [reply]
> Which operating system are you developing on? I ask this because running things out of a docker container is known for being slow

Unfortunately, I am running on a Windows machine which is probably why it is running so slowly.

> A container restart shouldn't be necessary though, you should be able to get away with just running make build in the container, then make dev.

I also tried this and I was still having the same issue. My teammate was able to have some luck by changing the homepage display text in home.esj but so far any changes I have made are not showing.

> You are very welcome! I know we've been steering people more towards posting on GitHub discussions, but for more tech-support type help where it can be useful to interact directly, check out the [pl-help](https://app.slack.com/client/T0PHEMPRV/C266KEH9A) channel on slack.

In regards to this would there be any chance our team might be able to join a voice call with one of your maintainers and maybe get some guidance? We've been struggling to get things to work for a bit now and would be very thankful to receive some live assistance.

**Reply by nwalters512** [reply]
> In regards to this would there be any chance our team might be able to join a voice call with one of your maintainers and maybe get some guidance? hours

We hold weekly open office hours, you're more than welcome to stop by! See the description of the `#pl-help` channel on Slack for up-to-date times and a link to the Zoom.

> I also tried this and I was still having the same issue. My teammate was able to have some luck by changing the homepage display text in home.esj but so far any changes I have made are not showing.

We can take a look at this in office hours, but in the meantime, could you post the exact commands you're running, files you're changing, etc? We can at least see if there appears to be anything off with what you're doing.

**Reply by RylanRCox** [reply]
> you can even develop natively on your system outside of a container (see instructions [here](https://github.com/PrairieLearn/PrairieLearn/blob/master/docs/installingNative.md)).

Attempting to try to run it natively is not working for me as when I run a "make deps" command I get a permission error. Which I'm unsure how to solve as I am running as admin on the admin account of my PC.

> We hold weekly open office hours, you're more than welcome to stop by!

That's very exciting to hear! We will most definitely be looking into that.

> We can at least see if there appears to be anything off with what you're doing.

So these are the commands I am putting into my terminal.

> docker run -it --rm -p 3000:3000 -w /PrairieLearn -v ${PWD}:/PrairieLearn prairielearn/prairielearn /bin/bash
> 
> Then in the container, we run:
> 
> make deps
> 
> make dev

I attempted to put a "make build" command in between "make deps" and "make dev" which did not change anything.

The change we're attempting to make is just something small to see if we can actually make changes to the container. We are changing PrairieLearn\apps\prairielearn\src\pages\home\home.ejs line 97 by making a small change to the string. When I refresh the page nothing changes.

Thanks again for your help and your incredibly fast response times.

**Reply by nwalters512** [reply]
> Attempting to try to run it natively is not working for me as when I run a "make deps" command I get a permission error.

If you're trying to run natively on Windows, you're probably going to have a bad time. I'd recommend WSL; I don't think anyone here can provide useful support for native development on Windows, and I'm not even sure if PrairieLearn supports that in the first place.

Re: the change you're trying to make, an easy test would be to `cat /PrairieLearn/apps/prairielearn/src/pages/home/home.ejs` inside the container to see if you see the expected change on line 97. There might be something strange happening with Docker on your system where the change you make in your editor isn't reflected inside the container?

**Comment by nwalters512** [comment]
Apologies about the state of the dev guide; it's fallen significantly out of date and basically needs to be rewritten from scratch. For up-to-date information on creating database migrations, see [`apps/prairielearn/src/migrations/README.md`](https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/src/migrations/README.md).
